### PR TITLE
Delete buckets and collections

### PR DESCRIPTION
**Pull request by leplatrem**
While I was add consistency in kinto.py, I realized that we could had `DELETE` endpoints to buckets and collections.

Of course, it deletes only the buckets/collections where the user has write permissions.

* [x] Implement missing methods
* [x] Rely on events instead of subclassing resource
* [x] Wait for https://github.com/mozilla-services/cliquet/pull/645 to be merged
* [x] Add tests for recursivity (should wipe everything under)
* [x] Update docs

### CODE REVIEW

**Comment by leplatrem** [comment]
This is now open for review/merge! 

@magopian, FYI this PR switches from inherited method in the resource class to event subscribers. 
@Natim r?

File: CHANGELOG.rst [hunk 1]
```diff
@@ -6,6 +6,14 @@ This document describes changes between each past release.
 1.12.0 (unreleased)
 ===================
 
+**Protocol**
+
+- Delete every (writable) buckets using ``DELETE /v1/buckets``
+- Delete every (writable) collections using ``DELETE /v1/buckets/<>/collections``
```
Comment (magopian) [inline CHANGELOG.rst]: Did you want a `<bucket id>` here instead of `<>`?

File: docs/api/1.x/index.rst [hunk 2]
```diff
@@ -43,7 +45,11 @@ Cheatsheet
 +----------+----------------------------------------------------------------------------------------------+---------------------------------------------------------+
 | **Collections**                                                                                                                                                   |
 +----------+----------------------------------------------------------------------------------------------+---------------------------------------------------------+
-| `POST`   | :ref:`/buckets/(bucket_id)/collections <collection-post>`                                    | :ref:`Create a collection <collection-put>`             |
```
Comment (magopian) [inline docs/api/1.x/index.rst]: good catch for the `collection-post` instead of `collection-put` here ;)

File: docs/api/index.rst [hunk 3]
```diff
@@ -15,7 +15,8 @@ Changelog
 ''''''''''''''''
 
 - Allow bucket to get arbitrary attributes.
-
+- Delete every (writable) buckets using ``DELETE /v1/buckets``
+- Delete every (writable) collections using ``DELETE /v1/buckets/<>/collections``
```
Comment (magopian) [inline docs/api/index.rst]: Same here, did you want to have `<bucket id>` instead of `<>`?

File: kinto/views/buckets.py [hunk 4]
```diff
@@ -23,12 +26,20 @@ def get_parent_id(self, request):
...
+            for_resources=('bucket',),
+            for_actions=(ACTIONS.DELETE,))
+def on_buckets_deleted(event):
+    """Some buckets were deleted, delete sub-resources.
+    """
+    storage = event.request.registry.storage
+
+    for change in event.impacted_records:
+        bucket = change['old']
+        parent_id = '/buckets/%s' % bucket['id']
```
Comment (magopian) [inline kinto/views/buckets.py]: nit: don't you use the better/faster/stronger `.format` for string interpolation?

File: kinto/views/groups.py [hunk 5]
```diff
@@ -1,6 +1,8 @@
 import colander
```
Comment (magopian) [inline kinto/views/groups.py]: Changes in this file aren't related to the PR, are they? You took the opportunity to change the group deletion to using the subscriber too?

**Comment by magopian** [comment]
From what I understand of the code I read, this looks good to me. If someone else with better knowledge of the codebase could double-check, that would be nice ;)

There's a test failure, from the load tests, is it related to what you were discussing with @Natim on irc about cliquet having timestamp unicity issues?

File: kinto/views/groups.py (hunk 5 above)
Comment (leplatrem) [inline kinto/views/groups.py]: Yes, you're right, I move everything from inheritance to events

File: kinto/views/buckets.py (hunk 4 above)
Comment (leplatrem) [inline kinto/views/buckets.py]: erf we should, but not only here...

http://makina-corpus.com/blog/metier/the-worlds-simplest-python-template-engine :)

**Pull request merged**